Reorganize and restructure Organization docs (#2758)
clerkinc/clerk-docs: docs/guides/organizations/get-started-tu/quickstart.mdx

---
title: Get started with Organizations and Next.js
description: Create and manage organizations in your Next.js app in this guided tutorial.
sdk: nextjs
---

Organizations let you group users with roles and permissions, enabling you to build multi-tenant B2B apps like Slack (workspaces), Linear (teams), or Vercel (projects) where users can switch between different team contexts.

> [!NOTE]
> Haven't setup Clerk auth yet? Head over to [Next.js Quickstart](/docs/nextjs/getting-started/quickstart)!

<Steps>

>   ## Add `<OrganizationSwitcher/>` to app

Reviewer: ```suggestion
  ## Add `<OrganizationSwitcher/>` to your app
```

  The `<OrganizationSwitcher />` component is the easiest way to let users create, switch between, and manage organizations. We recommend you place it in your app's header or navigation so it's always accessible. For example:

>   ```tsx {{ filename: 'app/layout.tsx', mark: [9, 39], fold: [[3, 8], [11, 28]] }}

Reviewer 2: I think the fold looks awkward on the imports, what do you think about leaving it expanded?

Before:
<img width="677" height="743" alt="Screenshot 2025-11-20 at 13 37 12" src="https://github.com/user-attachments/assets/33dbfc13-bf17-47ec-9c43-d919a186c4de" />

After:
<img width="677" height="743" alt="Screenshot 2025-11-20 at 13 37 08" src="https://github.com/user-attachments/assets/7a7e7a14-381b-4663-a2f3-a767bbca4921" />

```suggestion
  ```tsx {{ filename: 'app/layout.tsx', mark: [9, 39], fold: [[11, 28]] }}
```

  import { Geist, Geist_Mono } from 'next/font/google'
  import './globals.css'

  const geistSans = Geist({
    variable: '--font-geist-sans',
    subsets: ['latin'],
  })

  const geistMono = Geist_Mono({
    variable: '--font-geist-mono',
    subsets: ['latin'],
  })

  export const metadata: Metadata = {
    title: 'Clerk Next.js Quickstart',
    description: 'Generated by create next app',
  }

  export default function RootLayout({
    children,
  }: Readonly<{
    children: React.ReactNode
  }>) {
    return (
      <ClerkProvider>
        <html lang="en">
          <body className={`${geistSans.variable} ${geistMono.variable} antialiased`}>
            <header className="flex justify-end items-center p-4 gap-4 h-16">
              <OrganizationSwitcher />
              <SignedOut>
                <SignInButton />
                <SignUpButton>
                  <button className="bg-[#6c47ff] text-white rounded-full font-medium text-sm sm:text-base h-10 sm:h-12 px-4 sm:px-5 cursor-pointer">
                    Sign Up
                  </button>
                </SignUpButton>
              </SignedOut>
              <SignedIn>
                <UserButton />

>               </SignedIn>

Reviewer 2: I don't know if it matters, but I'd probably put the `<OrganizationSwitcher />` within the `<SignedIn>` component. Maybe it doesn't even show up unless the user is authenticated, so it doesn't matter rendering-wise, but just for best coding practices

```suggestion
              <SignedOut>
                <SignInButton />
                <SignUpButton>
                  <button className="bg-[#6c47ff] text-white rounded-full font-medium text-sm sm:text-base h-10 sm:h-12 px-4 sm:px-5 cursor-pointer">
                    Sign Up
                  </button>
                </SignUpButton>
              </SignedOut>
              <SignedIn>
                <UserButton />
                <OrganizationSwitcher />
              </SignedIn>
```

Author: it doesn't render if the user isnt authed

the reason why i wanted to put the OrgSwitcher outside of signed in is that the orgswitcher doesn't render initially (so clerk wont prompt folks to enable organizations)

when using orgs, we recommend b2b saas apps (our icp) to disable personal accounts. so the action of creating a user (that's not associated with an org), enabling org... and having an orphaned user... disrupts that mental model

can pair and show you how this works

    const { organization } = useOrganization()
    const { userMemberships } = useOrganizationList({
      userMemberships: true,
    })

    return (
      <div className="p-8">
        <h1 className="text-2xl font-bold mb-4">
          Welcome to the <strong>{organization?.name}</strong> organization
        </h1>
        <p className="mb-6">
          Your role in this organization:{' '}
          <strong>
            {
              userMemberships?.data?.find(
                (membership) => membership.organization.id === organization?.id,
              )?.role
            }
          </strong>
        </p>
      </div>
    )
  }
  ```

  ## Visit your app

  Run your project with the following command:

  ```npm
  npm run dev
  ```

  Visit your app locally at [`localhost:3000`](http://localhost:3000).

  ### Enable organization

  Since this app uses Organization hooks and components, you must enable the Organizations feature.

  Click on "Enable Organizations" when prompted.

  ### Create first user and organization

  You must sign in to use organizations. When prompted, click on "Sign in to continue". Then, authenticate to create your first user.

>   Since you selected disabled personal accounts when you enabled organizations, every user must be in at least one organization. Clerk will prompt you to create an organization for your user.

Reviewer 2: ```suggestion
  You must sign in to use organizations. When prompted, select **Sign in to continue**. Then, authenticate to create your first user.

  Since you selected disabled personal accounts when you enabled organizations, every user must be in at least one organization. Clerk will prompt you to create an organization for your user.
```

Reviewer 2: I feel like this is a good opportunity to mention session tasks here. like just something simple like a reference link:

Since you selected disabled personal accounts when you enabled organizations, every must be in at least one organization. [Clerk will prompt you](/docs/guides/configure/session-tasks) to create an organization for your user.

or we could even do a tooltip instead of the link ^
and the tooltip can say:
This happens because the `choose-organization` session task must be completed before the user's authentication process can be considered complete. Session tasks are pending requirements that users must complete after authentication. Learn more about [session tasks](/docs/guides/configure/session-tasks).

  ## Create and switch organization

  Use `<OrganizationSwitcher />` to create, switch between, and manage organizations.

  1. Click on the `<OrganizationSwitcher />` component, then **Create an organization**.

  1. Enter `Acme Corp` as the organization name.

  1. Invite users to your organization and select their role.

  ## Protect routes by organization and roles

>   You can protect routes based on organization membership, roles, and permissions. The `/protected` page checks if you are an admin in the `Acme Corp` organization:

Reviewer 2: ```suggestion
  You can protect content and even entire routes based on organization membership, roles, and permissions by performing [authorization checks](!authorization-check).

      In the following example, the page is protected from unauthenticated users, users that don't have the `org:admin` role, and users that are not in the `Acme Corp` organization.

      - The [`Auth`](/docs/reference/backend/types/auth-object) object is used to access the `isSignedIn` and `has()` properties.
      - The `isSignedIn` property is used to check if the user is signed in.
      - The `has()` helper is used to check if the user has the `org:admin` role.
      - The [`useOrganization()`](/docs/reference/hooks/use-organization) hook is used to access the organization data.
      - The organization name is checked to ensure it matches the required organization name. If a user is not in the required organization, the page will display a message and the [`<OrganizationSwitcher />`](/docs/reference/components/organization/organization-switcher) component will be rendered to allow the user to switch to the required organization.
```

or we can keep it simpler and rely on code comments to explain things (but the con there is that there aren't any reference links):

```suggestion
  You can protect content and even entire routes based on organization membership, roles, and permissions by performing [authorization checks](!authorization-check).

      In the following example, the page is protected from unauthenticated users, users that don't have the `org:admin` role, and users that are not in the `Acme Corp` organization.
```

  ## It's time to build your B2B SaaS!

  You've added Clerk organizations to your Next.js app 🎉. Ready to scale to enterprise customers?

  To make configuration changes to your Clerk development instance, claim the Clerk keys that were generated for you by selecting **Claim your application** in the bottom right of your app. This will associate the application with your Clerk account.

  - **Control access** with [custom roles and permissions](/docs/guides/organizations/control-access/roles-and-permissions) define granular permissions for different user types within organizations.

  - **Onboard entire companies** with [verified domains](/docs/guides/organizations/add-members/verified-domains) automatically invite users with approved email domains (e.g., `@company.com`) to join organizations without manual invitations.

  - **Enable enterprise SSO** with [SAML and OIDC](/docs/guides/organizations/add-members/sso) let customers authenticate through their identity provider (Okta, Entra ID, Google Workspace) with unlimited connections, no per-connection fees.
</Steps>

> ## Next steps

Author: i think we can just get rid of this section. you're right, it's more of a duplicate of the ones above
